Add hive boost to climate and water_heater

KJonline opened the pull request:
## Description:
Add the ability to call the boost function for the hive hot water. it will also allow a time period to be set for the climate boost by implementing a hive boost service.

**Related issue (if applicable):** fixes #25536

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10407


## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [X] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

Commit:
Start the Boost work

Commit:
Start the Boost work

Commit:
Add services.yaml

Commit:
Add services.yaml

Commit:
Added Services #2

Commit:
Added Services #2

Commit:
Merge branch 'feature/add_boost' of github.com:KJonline/home-assistant into feature/add_boost

Commit:
Working Services

Commit:
Merge branch 'dev' into feature/add_boost

* dev: (42 commits)
  Remove default host for Plex config (#26583)
  Add Obihai integration (#26537)
  Add switch platform to iaqualink integration (#26545)
  Fix GTFS sensor wrong timezone (#26580)
  Make uk_transport sensor timezone/DST aware (#26577)
  Updated frontend to 20190911.0 (#26578)
  bump dependencies (#26576)
  Bump UPC connect / fix auth issue (#26570)
  Cleanup stale script stuff (#26573)
  Improve startup of devcontainer (#26572)
  Ignore test output
  [ci skip] Translation update
  deCONZ device automations (#26366)
  Nuki less strict (#26542)
  osram cluster (#26555)
  fix events for smartthings acceleration cluster (#26557)
  Fix tests
  Refactor Cast (#26550)
  Add sensor platform to iaqualink component (#26544)
  Bump geniushub-client to 0.6.13 (#26554)
  ...

Rendili committed:
pyhiveapi to 0.2.19

Commit:
Update Libary to 0.2.19

Commit:
Merge branch 'feature/add_boost' of github.com:KJonline/home-assistant into feature/add_boost

* 'feature/add_boost' of github.com:KJonline/home-assistant:
  Update Libary to 0.2.19

Commit:
Merge branch 'dev' into feature/add_boost

* dev: (38 commits)
  Fix Environment Canada weather forecast, retain icon_code sensor (#26646)
  Rename MockToggleDevice to MockToggleEntity (#26644)
  Refactor nzbget to support future platform changes (#26462)
  deCONZ - Remove mechanisms to import a configuration from configuration.yaml (#26648)
  deCONZ - battery sensor instead of battery attribute (#26591)
  Add built in weather to Homematic IP Cloud (#26642)
  Move deCONZ services to their own file (#26645)
  Add group attribute to Homematic IP Cloud (#26618)
  Add iaqualink binary sensor and unique_id (#26616)
  zha ZCL color loop effect (#26549)
  [ci skip] Translation update
  deCONZ -  create deconz_events through sensor platform (#26592)
  Update azure-pipelines-wheels.yml for Azure Pipelines
  Update azure-pipelines-wheels.yml
  Refactor Bluetooth Tracker to async (#26614)
  Fix Typo (#26612)
  [ci skip] Translation update
  Disable Watson TTS Telemetry (#26253)
  Improve bluetooth tracker device code (#26067)
  Bump zigpy-zigate to 0.3.1 (#26600)
  ...

Rendili committed:
Update Water_heater boost

Commit:
Added Async hass add function

Commit:
Update Services

Commit:
Merge branch 'dev' into feature/add_boost

* dev: (92 commits)
  Bump pynws version to 0.8.1 (#26770)
  Bump HAP-python to 2.6.0 for homekit (#26783)
  [ci skip] Translation update
  Add integration scaffolding script (#26777)
  Bump simplisafe-python to 5.0.1 (#26775)
  Bump aiowwlln to 2.0.2 (#26769)
  ZHA siren and warning device support (#26046)
  Add transport data from maps.yandex.ru api (#26766)
  Type hint additions (#26765)
  Bump openwrt-luci-rpc to version 1.1.1 (#26759)
  Revert "Add transport data from maps.yandex.ru api (#26252)" (#26762)
  [ci skip] Translation update
  Updated frontend to 20190919.0
  deCONZ improve gateway tests (#26709)
  Add transport data from maps.yandex.ru api (#26252)
  Update codeowners (#26733)
  Bump influxdb to 5.2.3 (#26743)
  Izone component (#24550)
  Bump restrictedpython to 5.0 (#26741)
  Bumped version to 0.99.1
  ...

Commit:
MERGED

MartinHjelmare commented on homeassistant/components/hive/climate.py:
```diff
@@ -28,6 +32,15 @@
+BOOST_HEATING_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_BOOST_MINUTES): cv.positive_int,
```
We want to use our `cv.time_period` for all time interval config or service options.
```py
vol.All(
    cv.time_period,
    cv.positive_timedelta,
    lambda td: td.total_seconds() // 60
)
```

KJonline replied:
Currently changing this now

KJonline replied:
Complete

KJonline commented on homeassistant/components/hive/binary_sensor.py:
```diff
@@ -65,6 +59,17 @@ def is_on(self):
         """Return true if the binary sensor is on."""
         return self.session.sensor.get_state(self.node_id, self.node_device_type)
 
+    async def async_added_to_hass(self):
```
Hi @MartinHjelmare  becuase i am trying to implement  the refresh system as per below, i dont think i will need this as i have left async_added_to_hass only in the platforms that need it (2 of them)

KJonline commented on homeassistant/components/hive/climate.py:
```diff
@@ -42,6 +55,19 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
+    def heating_boost(service):
+        """Handle the service call."""
+        session = hass.data.get(DATA_HIVE)
```
@MartinHjelmare  - This has been updated.

KJonline commented on homeassistant/components/hive/climate.py:
```diff
@@ -42,6 +55,19 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+    def heating_boost(service):
+        """Handle the service call."""
+        session = hass.data.get(DATA_HIVE)
+        entity = session.entity_lookup[service.data.get(ATTR_ENTITY_ID)]
```
@MartinHjelmare  - This has been updated.

KJonline commented on homeassistant/components/hive/climate.py:
```diff
@@ -42,6 +55,19 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
     add_entities([climate])
 
+    def heating_boost(service):
```
i didnt want to add the service handlers and registration to the componet as the services should not registered if the user does not have heating or hot water. E.G they only had hive lights and a hub, the user should see a heating or water heater service.

MartinHjelmare replied:
I think we should sort the node ids per ha_type so that we only load platform once per ha_type. Then we can also register the corresponding service,from the component, when iterating the sorted dict. Probably use a dict with lists of devices.

KJonline replied:
i believe thats what we currently have now we just drill down and initialise each device in the component instead of the platform. how would you suggest we did it change the load platform to itertate through each device adding an entity in the setup platform function?

DEVICETYPES = {
    "binary_sensor": "device_list_binary_sensor",
    "climate": "device_list_climate",
    "water_heater": "device_list_water_heater",
    "light": "device_list_light",
    "switch": "device_list_plug",
    "sensor": "device_list_sensor",
}

MartinHjelmare replied:
Something like this but simplified:
https://github.com/home-assistant/home-assistant/blob/5914475fe5f97d08e9d721cbf328727df2f3af5e/homeassistant/components/mysensors/gateway.py#L186-L196

MartinHjelmare replied:
Pass the list of devices to `load_platform`. In each platform, iterate this list to create all the entities to add.

KJonline commented on homeassistant/components/hive/water_heater.py:
```diff
@@ -32,6 +40,22 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
     add_entities([water_heater])
 
+    def hotwater_boost(service):
```
same as above

Commit:
Reviewed Changes

KJonline commented on homeassistant/components/hive/binary_sensor.py:
```diff
@@ -65,6 +59,17 @@ def is_on(self):
+    async def async_added_to_hass(self):
+        """When entity is added to Home Assistant."""
+        await super().async_added_to_hass()
+        self.session.entities.append(self)
```
@MartinHjelmare i am currently trying to implement this but i am getting issues as hive is synchronous calling asynchronous code and it is causing problems.

homeassistant/components/hive/__init__.py", line 93, in wrapper
    await func(self, *args, **kwargs)
TypeError: object NoneType can't be used in 'await' expression

KJonline replied:
I seem to have got around the above error by changing the wrapper function from a Async one to be just sync. Will this cause any issue as it’s calling the async dispatcher within it?

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -79,3 +82,15 @@ def setup(hass, config):
+    """Force update all entities after state change."""
+
+    @wraps(func)
+    async def wrapper(self, *args, **kwargs):
```
Make this a regular function, not a coroutine.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -79,3 +82,15 @@ def setup(hass, config):
+    @wraps(func)
+    async def wrapper(self, *args, **kwargs):
+        """Call decorated function and send update signal to all entities."""
+        await func(self, *args, **kwargs)
```
Just call `func` don't `await` it.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -79,3 +82,15 @@ def setup(hass, config):
+    async def wrapper(self, *args, **kwargs):
+        """Call decorated function and send update signal to all entities."""
+        await func(self, *args, **kwargs)
+        async_dispatcher_send(self.hass, DOMAIN)
```
Use `dispatcher_send`.

KJonline replied:
Completed

MartinHjelmare commented on homeassistant/components/hive/climate.py:
```diff
@@ -177,8 +191,10 @@ def set_preset_mode(self, preset_mode) -> None:
+    async def async_added_to_hass(self):
+        """When entity is added to Home Assistant."""
+        await super().async_added_to_hass()
+        self.session.entity_lookup.update({self.entity_id: self.node_id})
```
```py
self.session.entity_lookup[self.entity_id] = self.node_id
```

MartinHjelmare replied:
We need to connect the signal, `DOMAIN`, to a callback update method here. I'm thinking we then need this in all platform entity classes, so a common base class would be good.

KJonline replied:
I have updated how it is added to the dict, but i am not sure what you mean by connect the signal and domain to a call back method.

MartinHjelmare replied:
This part:
https://github.com/home-assistant/home-assistant/blob/0e157856027d854e4183e00d91c830bfcdad877f/homeassistant/components/iaqualink/__init__.py#L192-L198

MartinHjelmare commented on homeassistant/components/hive/climate.py:
```diff
@@ -42,6 +57,19 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+    def heating_boost(service):
+        """Handle the service call."""
+        node_id = HiveSession.entity_lookup[service.data.get(ATTR_ENTITY_ID)]
+        minutes = service.data.get(ATTR_TIME_PERIOD)
```
```py
minutes = service.data[ATTR_TIME_PERIOD]
```

KJonline replied:
Complete

MartinHjelmare commented on homeassistant/components/hive/climate.py:
```diff
@@ -42,6 +57,19 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+    @refresh_system
+    def heating_boost(service):
+        """Handle the service call."""
+        node_id = HiveSession.entity_lookup[service.data.get(ATTR_ENTITY_ID)]
```
We don't know that the entity_id will be correct, right?
```py
node_id = HiveSession.entity_lookup.get(service.data[ATTR_ENTITY_ID])
if not node_id:
    # log or raise error
    return
```

KJonline replied:
Done

MartinHjelmare commented on homeassistant/components/hive/climate.py:
```diff
@@ -42,6 +57,19 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        """Handle the service call."""
+        node_id = HiveSession.entity_lookup[service.data.get(ATTR_ENTITY_ID)]
+        minutes = service.data.get(ATTR_TIME_PERIOD)
+        temperature = float(service.data.get(ATTR_TEMPERATURE, 25))
```
Put the default in the service schema and coerce to float in the schema.

KJonline replied:
I have put the default in the service schema but i am not sure how to represent a float with config validation. the only option i could find is small_float but that has a maximum of 1

MartinHjelmare replied:
Use `vol.Coerce(float)`.

Commit:
Fixed Refresh System

Commit:
Review 2

Commit:
Moved device iteration to the platform

KJonline commented:
Hi @MartinHjelmare  i have updated and moved the service registration and the service handler to the component level now. i have also updated the component to pass through a list for each HA type and in turn within the platform setup iterating through that list adding the entity.

I am struggling with the dispatch_connect and the handle callback. I have tried implementing it as a base class and passing it through, i have even tried implementing it directly into the platform. For some reason which ever way i try an achieve it the platform i add it to will freeze during setup and HA will stay frozen and not report any errors. Any ideas?

Commit:
update

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -54,16 +83,45 @@ class HiveSession:
+            return
+
+        minutes = service.data[ATTR_TIME_PERIOD]
+        temperature = service.data.get(ATTR_TEMPERATURE)
```
Please use `dict[key]` for required schema keys and keys with default schema values.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -51,16 +83,45 @@ class HiveSession:
+            # log or raise error
+            _LOGGER.error("Cannot boost entity id entered.")
+            return
+        minutes = service.data.get(ATTR_TIME_PERIOD)
```
As above.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -51,16 +83,45 @@ class HiveSession:
+        node_id = HiveSession.entity_lookup.get(service.data[ATTR_ENTITY_ID])
+        if not node_id:
+            # log or raise error
+            _LOGGER.error("Cannot boost entity id entered.")
```
We don't end logging messages with period.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -73,9 +134,34 @@ def setup(hass, config):
-                for hivedevice in devices:
-                    load_platform(hass, ha_type, DOMAIN, hivedevice, config)
+    for ha_type in DEVICETYPES:
+        devicelist = devices.get(DEVICETYPES[ha_type], None)
```
`None` is the default value returned if the key is missing in the dict.

MartinHjelmare commented on homeassistant/components/hive/binary_sensor.py:
```diff
@@ -10,9 +10,10 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-    add_entities([HiveBinarySensorEntity(session, discovery_info)])
+    session = hass.data.get(DATA_HIVE)
+    for device in discovery_info:
+        add_entities([HiveBinarySensorEntity(session, device)])
```
Please collect all entities in a list before calling `add_entities` to only call `add_entities` once.

MartinHjelmare commented on homeassistant/components/hive/water_heater.py:
```diff
@@ -88,18 +78,21 @@ def operation_list(self):
+        """When entity is added to Home Assistant."""
+        await super().async_added_to_hass()
+        self.session.entity_lookup[self.entity_id] = self.node_id
+        dispatcher_connect(self.hass, DOMAIN, self._update_callback)
```
Use `async_dispatcher_connect`. We're in async context here since we're in a coroutine.

MartinHjelmare commented on homeassistant/components/hive/water_heater.py:
```diff
@@ -88,18 +78,21 @@ def operation_list(self):
+
+    @callback
+    def _update_callback(self) -> None:
+        self.schedule_update_ha_state()
```
Use `async_schedule_update_ha_state`. We're in async context since we're decorated with `@callback`.

Commit:
Updates #2

Commit:
Review#3 New Base Class

KJonline commented:
Hi @MartinHjelmare, I have amended post your recent comments and updated the files passing a list through to Add Entities and introducing a base class for the callback, along with the other changes.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -165,3 +154,29 @@ def wrapper(self, *args, **kwargs):
     return wrapper
+
+
+class HiveSession(Entity):
```
The hive session shouldn't inherit from `Entity`.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -73,9 +123,60 @@ def setup(hass, config):
+    weather = None
+    attributes = None
+
+    async def async_added_to_hass(self):
```
We should have a separate base class that inherits from `Entity` for the hive entities. The session class should be separate from that.

MartinHjelmare commented on homeassistant/components/hive/binary_sensor.py:
```diff
@@ -10,12 +10,15 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
 
-class HiveBinarySensorEntity(BinarySensorDevice):
+class HiveBinarySensorEntity(HiveSession, BinarySensorDevice):
```
We should inherit from a new hive entity class, not from the `HiveSession`.

MartinHjelmare commented on homeassistant/components/hive/light.py:
```diff
@@ -156,7 +149,8 @@ def supported_features(self):
-        self.session.core.update_data(self.node_id)
-        self.attributes = self.session.attributes.state_attributes(self.node_id)
+
+def update(self):
```
This looks like it has been outdented too far.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -73,9 +123,60 @@ def setup(hass, config):
+
+    async def async_added_to_hass(self):
+        """When entity is added to Home Assistant."""
+        await super().async_added_to_hass()
```
We don't need to call super. There's no parent method to call.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -73,9 +123,60 @@ def setup(hass, config):
+        await super().async_added_to_hass()
+        async_dispatcher_connect(self.hass, DOMAIN, self._update_callback)
+        if self.device_type in SERVICES:
+            self.session.entity_lookup[self.entity_id] = self.node_id
```
We need an init method that initializes the instance attributes that we need to access here.

Commit:
Review #5

KJonline commented:
Hi @MartinHjelmare, i have committed the latest changes following your lest review.

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -156,22 +170,18 @@ def wrapper(self, *args, **kwargs):
-    switch = None
-    weather = None
-    attributes = None
+    def __init__(self):
```
Accept the corresponding parameters for the instance attributes here. Then we don't need to set the attributes both in parent and in child class, but just in the parent.

```py
def __init__(self, session, hive_device):
    self.node_id = hive_device["Hive_NodeID"]
    self.node_name = hive_device["Hive_NodeName"]
    self.device_type = hive_device["HA_DeviceType"]
    self.node_device_type = hive_device["Hive_DeviceType"]
    self.session = session
    self.attributes = {}
    self._unique_id = f"{self.node_id}-{self.device_type}"
```

```suggestion
    def __init__(self, session, hive_device):
```

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -73,9 +137,56 @@ def setup(hass, config):
+    """Initiate Hive Base Class."""
+
+    def __init__(self):
+        """Initialise the varibale."""
```
```suggestion
        """Initialize the instance."""
```

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -73,9 +137,56 @@ def setup(hass, config):
+
+    def __init__(self):
+        """Initialise the varibale."""
+        self.entity_id = None
```
This is already defined in the `Entity` class.
```suggestion
```

Commit:
Merge branch 'dev' into feature/add_boost

* dev: (87 commits)
  Add ecobee services to create and delete vacations (#26923)
  Centralize rainbird config and add binary sensor platform (#26393)
  Add config flow to transmission (#26434)
  Add Plex config options support (#26870)
  Bump pyobihai, add unique ID and availability (#26922)
  Add mysensors codeowner (#26917)
  [ci skip] Translation update
  Add MySensors ACK (#26894)
  Remove lamps and groups from ha when removed from Hue (#26881)
  Add config flow to ecobee (#26634)
  deCONZ - Increase bridge discovery robustness in config flow (#26911)
  Add call direction sensor for Obihai (#26867)
  Bumped version to 0.99.3
  HM-CC-TC was not recognized (#26623)
  Add google_assistant alarm_control_panel (#26249)
  deCONZ - Improve ssdp discovery by storing uuid in config entry (#26882)
  Fix missing whitespace around arithmetic operator (#26908)
  Fix bed_activity history chart of the Xiaomi Aqara vibration sensor (#26875)
  Add voltage attribute to Xiaomi Aqara devices (#26876)
  Bump ndms2-client to 0.0.9 (#26899)
  ...

MartinHjelmare commented on homeassistant/components/hive/binary_sensor.py:
```diff
@@ -26,9 +29,7 @@ def __init__(self, hivesession, hivedevice):
         self.node_device_type = hivedevice["Hive_DeviceType"]
```
Here call the init method of the parent class.
```py
super().__init__(hive_session, hive_device)
```

Commit:
Update homeassistant/components/hive/__init__.py

Co-Authored-By: Martin Hjelmare <marhje52@kth.se>

Commit:
Update homeassistant/components/hive/__init__.py

Co-Authored-By: Martin Hjelmare <marhje52@kth.se>

Commit:
Update homeassistant/components/hive/__init__.py

Co-Authored-By: Martin Hjelmare <marhje52@kth.se>

Commit:
Review 6

KJonline commented:
Hi @MartinHjelmare, i have updated moving the variables to the parent class.  i have had to leave one variable in the light child class as it has an extra piece of data which is only available if the device is  light.

Commit:
Review 7

MartinHjelmare commented on homeassistant/components/hive/__init__.py:
```diff
@@ -73,9 +137,59 @@ def setup(hass, config):
+        self.node_id = hive_device["Hive_NodeID"]
+        self.node_name = hive_device["Hive_NodeName"]
+        self.device_type = hive_device["HA_DeviceType"]
+        self.node_device_type = hive_device["Hive_DeviceType"]
```
This attribute seems only used in the binary sensor platform. We can move it to that class.

KJonline replied:
@MartinHjelmare , we do populate that data for all the other devices/platforms. Is it worth leaving it in the parent class in case that data is used in the future with other platforms other than binary_sensor?

MartinHjelmare replied:
Ok, let's keep it.

KJonline commented:
Hi @MartinHjelmare, it looks like the pylint check is failing on the platform classes that call the parent class but don’t do anything else within the initialisation process e.g binary_sensor where as the climate and light platform which calls the parent and assigns a variable in the child class aren’t being flagged by pylint.  Any ideas?

MartinHjelmare commented:
In the child classes that doesn't do anything but call the parent class init, we can remove the init method in the child class.

MartinHjelmare commented on homeassistant/components/hive/binary_sensor.py:
```diff
@@ -10,25 +10,20 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     """Representation of a Hive binary sensor."""
 
-    def __init__(self, hivesession, hivedevice):
+    def __init__(self, hive_session, hive_device):
```
Remove this.

MartinHjelmare commented on homeassistant/components/hive/sensor.py:
```diff
@@ -19,27 +19,21 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     """Hive Sensor Entity."""
 
-    def __init__(self, hivesession, hivedevice):
+    def __init__(self, hive_session, hive_device):
```
Remove this.

MartinHjelmare commented on homeassistant/components/hive/switch.py:
```diff
@@ -1,31 +1,27 @@
     """Hive Active Plug."""
 
-    def __init__(self, hivesession, hivedevice):
+    def __init__(self, hive_session, hive_device):
```
Remove this.

MartinHjelmare commented on homeassistant/components/hive/water_heater.py:
```diff
@@ -1,50 +1,39 @@
     """Hive Water Heater Device."""
 
-    def __init__(self, hivesession, hivedevice):
+    def __init__(self, hive_session, hive_device):
```
Remove this.

KJonline commented:
Hi @MartinHjelmare, apoligies i am not familiar with using super(). but when i remove the __init__ method from the child classes i get an error advising

` File "/Users/kholejones/Git/home-assistant/homeassistant/components/hive/sensor.py", line 31, in <module>
    class HiveSensorEntity(HiveEntity, Entity):
  File "/Users/kholejones/Git/home-assistant/homeassistant/components/hive/sensor.py", line 34, in HiveSensorEntity
    super().__init__(hive_session, hive_device)`


I have updated the classes to be 

`class HiveSensorEntity(HiveEntity, Entity):
    """Hive Sensor Entity."""

    super().__init__(hive_session, hive_device)

    @property
    def unique_id(self):
        """Return unique ID of entity."""
        return self._unique_id`

MartinHjelmare commented:
Remove the whole init method including the call to `super()`.

Commit:
Removed Child classes to inhertit from the parent

KJonline commented:
Hi @MartinHjelmare, that’s seems to have passed all the checks now. I just wanted to say thanks for all your help with this. I started learning python with this component and this had been a massive learning curve. Thanks for your patience with me/this and there is still a lot for me to learn but I am looking forward to it.

MartinHjelmare reviewed (approved):
Great!

frenck commented on homeassistant/components/hive/__init__.py:
```diff
@@ -1,17 +1,32 @@
 DOMAIN = "hive"
 DATA_HIVE = "data_hive"
+SERVICES = ["Heating", "HotWater"]
+SERVICE_BOOST_HOTWATER = "boost_hotwater"
```
Noticed during the review of the documentation: `hotwater` is not one word. So I expected it to be `hot_water` ?
